### ISSUE TITLE
Implementing hyperparameters exchange  for declearn optimizers

### ISSUE THREAD

**Issue by srcansiz** [fedbiomed/fedbiomed#467]
In GitLab by @ybouilla on Jan 19, 2023, 10:30

This task is about finding out an appropriate way to deal with declearn optimizers hyperparameters sent through the network.
In declearn, optimizer hyperparameters are sent through `collect_aux_var` method, and received / parsed through `process_aux_var` 

In FedBioMed, there are 2 ways to sent hyperparameters (more generally speaking information) between Nodes and Researcher:
1. through MQTT message broker (JSON format): this is for small messages/ piece of information, like sharing a few values 
2. through file system exchange: this is for large messages / piece of information, such as model parameters or scaffold 's correction state

The decision of how to send messages in FedBioMed should be defined by the developer somehow - it should be hard-coded, and not just based on the size of the piece of information exchanged

The goal of this task is to:
 - [x] decide which object has the responsibility to send such information (Aggregator class in Fed-BioMed?)
 - [x] find a way to detail in Fed-BioMed how parameters required by declearn optimizers should be sent - it could be a Enum class that specify which parameters are required, that is consulted before sending messages for instance
 - [x] create interface / wrapper to use declearn 's optimizers `collect_aux_var` and `process_aux_var`  methods
  - ~~[ ] make hyper-parameters exchange compatible with Secure Aggregation
Assignee can detail in this issue proposals that can be discussed and acknowledged by Fed-BioMed team~~ replaced by #837

**Comment by ybouilla** [comment]
**About secure aggregation**:  auxiliary variables of declearn are exchanged decrypted, with may be an issue if user is using `Secure Aggregation`   or `DP`:

- for `Secure Aggregation`, a idea could be to create a declearn Scaffold warpper, that passes `Aggregator` as an argument (deltas sent back by `Nodes` to `Researcher` will be averaged)
- for `DP`, we should state clearly that for now, DP is only used on model weights (maybe add a warning message). 

For `Secure Aggregation`, if we go for a Scaffold-declearn warpper in Fed-BioMed, then we will have to choose with the normal way in declearn to use scaffold:

```
Optimizer(lr=.01, modules=[ScaffoldClientModule()])  # or the other way using strings
```
Will this be raising an error or work anyway (with/without encrypting auxiliary variables)

**Comment by ybouilla** [comment]
Work for making Optimizers that rely on auxiliary variables that are compatible with secagg :

1. bout keys :
 1.a . Do we need specific keys for model Weights and another one for auxiliary variables?
 it is not needed
 a.b. Should weights and auxiliary variables being processed separately or jointly?
It could be done separately using same key. The only issue is how to scale back auxiliary variables after secagg? 

2. how to use Scaffold with declearn?
several options:
option 1: ask Declearn to change their API, and accept an aggregated term as optional argument
option 2: compute aggregate ie the average auxiliary variable and pass same auxiliary variable for each client. Declearn will compute the mean of the same value, so it gives the correct  value.
` 1. x = mean(nodes_aux_var_{0<i<card(nodes)})
2.  for i in card(nodes):
              nodes_aux_var_i = x
3.  Optimizer.set_aux(nodes_aux_var)
 `

**Comment by ybouilla** [comment]
**About synchronicity**:
I think we need a method in `Experiment` that check correct synchronicity between Nodes Optimizer and Researcher' s one. This method can be in `Experiment` class (can be called `check_optimizer_synchronicity` for instance)

**Comment by ybouilla** [comment]
**Proposal**

a scaffold wrapper (not sure if compatible with declearn's registration tool)

```

class SecAggScaffold(OptiModule):
      def __init__(self, sec_agg_active:bool, *args, **kwargs):
           # 
          super().__init__(*args, **kwargs)
          self.aggregator = FedAvg()

     def set_aux(self, aux_var: dict):
          # ...
          # do aggregation here and compute the mean
          super().set_aux(aux_var)
```

**Comment by pandrey-fr** [comment]
**Regarding DP**:

There should not be an issue with Scaffold (at least there is no in DecLearn's DP implementation) provided the clipping and noise addition happen _before_ the gradients pass through the `ScaffoldClientModule` plug-in. The Scaffold update that clients compute and that needs aggregating is the average of the gradients received at each step during a round: as a consequence, if the inputs were already clipped and noised, this constitutes a post-processing step that does not alter the obtained DP guarantees.

In DecLearn, we achieve this by (a) clipping sample-wise gradients as part of their computation (b) introducing a `GaussianNoiseModule` plug-in as the first element of the `Optimizer` modules' pipeline. (a) is most probably handled similarly in Fed-BioMed; I do not know how (b) is handled however.

**Regarding SecAgg**:

The current implementation of `Scaffold` in declearn has the server-side plug-in hold some individual information and perform the aggregation computations. The simplest solution to enable using SecAgg is therefore to implement an alternative pair of plug-ins that have a different management of information and computations between the clients and server, enabling the use of SecAgg. This can be done either as part of DecLearn or of Fed-BioMed, using the type-registration utils to make them compatible with the entire DecLearn machinery. I think it should constitute a distinct issue, and tackled soon - I am eager to work on it during the current sprint. In the short run (declearn 2.3+), both versions of Scaffold would co-exist. In the longer run (declearn 3), we would deprecate the current one in favor of keeping only the SecAgg-compatible one.

**Comment by pandrey-fr** [comment]
Regarding the core of this issue: I started implementing on feature branch 467.

Advancement tracker:
- [x] Send from Researcher to Nodes
- [ ] Collect & Process on Node side
- [ ] Send from Node to Researcher
- [ ] Collect & Process on Researcher side

Once all steps are tackled I will open a draft PR, and write (hopefully, finalize) associated tests.

**Comment by ybouilla** [comment]
> Note that I only have sample-level DP-SGD in mind (when we want to be protected from membership inference attacks of any individual data sample); the issue might be different for other kinds of privacy settings.

I think we cover only that too in Fed-BioMed

**Comment by pandrey-fr** [comment]
> That is very nice if you could update declearn to make it work with FBM!!! I don't see how type-registration feature would be helpful here though?

Type-registration is useful insofar as it enables extending declearn. I.e. if you wrote the new pair of `declearn.optimizer.modules.OptiModule` subclasses as part of the Fed-BioMed source code, it would still get registered with declearn on import, so that if you used `declearn.optimizer.list_optim_modules()` they would appear as part of it, and, more importantly, they could be properly (de)serialized into/from configuration dicts that refer to them by a unique string identifier. In other words, while in this case I am very happy to do the implementation as part of declearn, it is _always_ possible for you to add new plug-ins directly as part of Fed-BioMed.

> By the way how should we handle another Optimizer that is not compatible with DP

To be honest, I don't know. To me it is up to the researcher to set up things that make sense (e.g. don't use DP + something that would break it), but we might come up with some ways to control that in the future, if we find that it is useful. That being said, it is not pure luck that Scaffold is compatible with DP: most optimizer modifications fall under the category of post-processing, and are therefore non-breaking wrt DP guarantees; you can think of all adaptive algorithms for instance (Adam, Adagrad, etc.).

**Comment by pandrey-fr** [comment]
OK, I have a basic version that should work (but does not tackle SecAgg). I still have to write/run associated unit and, more importantly, integration tests, which I will do either on Friday or next week.

As usual when working on Experiment (and, to a lesser extent, Job/Node/Round), I spent a lot of time debugging the existing monkey-patched-as-hell unit "tests". I even found a place where the code goes through useless convolutions _just_ for the tests to pass - you will find a note in the code regarding that. This is out of topic for this issue, and I know most of the team is already aware of it, but this code really should undergo some major refactoring at a point of time when no new feature is being implemented...
